File: BinarySearchTree/AVLTree.cpp
#include <iostream>

using namespace std;

template <typename T> struct Node {
    T data;
    Node* left;
    Node* right;
    int height;

    Node (T data)
    {
        this->data = data;
        this->left = nullptr;
        this->right = nullptr;
        this->height = 1;
    }

    Node ()
    {   
        this->data = T{};
        this->left = nullptr;
        this->right = nullptr;
        this->height = 0;
    }
};


template <typename T> class AVLTree
{
    private:
        int size;
        Node<T>* root;

        int height(Node<T> *node)
        {
            if(node == nullptr)
                return 0;
            return node->height;
        }

        int get_balance(Node<T> *node)
        {
            if(node == nullptr)
                return 0;
            
            return height(node->left) - height(node->right);
        }

        Node<T> *right_rotate(Node<T> *A)
        {
            Node<T> *B = A->left;
            Node<T> *Br = B->right;

            A->left = Br;
            B->right = A;

            A->height = 1 + max(height(A->left), height(A->right));            
            B->height = 1 + max(height(B->left), height(B->right));       

            return B;     
        }

        Node<T> *left_rotate(Node<T> *A)
        {
            Node<T> *B = A->right;
            Node<T> *Bl = B->left;

            A->right = Bl;
            B->left = A;

            A->height = 1 + max(height(A->left), height(A->right));            
            B->height = 1 + max(height(B->left), height(B->right));       

            return B;
        }

        Node<T> *find_left_successor(Node<T> *node)
        {
            Node<T> *temp = node;

            while(temp && temp->right)
                temp = temp->right;
            
            return temp;
        }

        Node<T> *find_right_successor(Node<T> *node)
        {
            Node<T> *temp = node;
            
            while(temp && temp->left)
                temp = temp->left;
            
            return temp;
        }

        Node<T>* insert(Node<T>* node, T data)
        {
            // Base Case
            if(node == nullptr) return new Node<T>(data);

            if(data > node->data)
                node->right = insert(node->right, data);
            else if(data < node->data)
                node->left = insert(node->left, data);
            else
                return node;
            
            // updates heights after insertion;
            // new inserted node will get 0, its parents will be incremented.
            node->height = 1 + max(height(node->left), height(node->right));

            int balance = get_balance(node);

            if(balance > 1 && data < node->left->data)
                return right_rotate(node);
            
            if(balance < -1 && data > node->right->data)
                return left_rotate(node);

            if(balance > 1 && data > node->left->data){
                node->left = left_rotate(node->left);
                return right_rotate(node);
            }

            if(balance < -1 && data < node->right->data){
                node->right = right_rotate(node->right);
                return left_rotate(node);
            }

            return node;

        }
        
        Node<T>* remove(Node<T>* node, T data)
        {
            if(node == nullptr) return node;

            if(data > node->data)
                node->right = remove(node->right, data);
            
            if(data < node->data)
                node->left = remove(node->left, data);
            else {
                // 4 cases!;
                Node<T>* temp = nullptr;

                if(node->right == nullptr)
                {
                    temp = node->left;
                    delete node;
                    return temp;
                }
                if(node->left == nullptr)
                {
                    temp = node->right;
                    delete node;
                    return temp;
                }

                if(node->right && node->left)
                {
                    Node<T> *successor = nullptr;
                    successor = find_right_successor(node->right);

                    // Uncomment these two commented lines below to use left successor as preference
                    // successor = find_left_successor(node->left);
                    
                    node->data = successor->data;
                    
                    // node->left = remove(node->left, successor->data);

                    // Comment this line if you want to use left successor
                    node->right = remove(node->right, successor->data);
                } 
            }
            
             node->height = height(node->left) + 
                   height(node->right) + 1;

            int balance = get_balance(node);

            if (balance > 1 && 
                get_balance(node->left) >= 0)
                return right_rotate(node);

            if (balance > 1 && 
                get_balance(root->left) < 0) {
                node->left = left_rotate(node->left);
                return right_rotate(node);
            }

            if (balance < -1 && 
                get_balance(node->right) <= 0)
                return left_rotate(node);


            if (balance < -1 && 
                get_balance(node->right) > 0) {
                node->right = right_rotate(node->right);
                return left_rotate(node);
            }


            return node;
        }

        void in_order_traversal(Node<T>* node)
        {
            if(node == nullptr) return;
            in_order_traversal(node->left);
            cout << node->data << " ";
            in_order_traversal(node->right);
        }

        void pre_order_traversal(Node<T>* node)
        {
            if(node == nullptr) return;
            cout << node->data << " ";
            pre_order_traversal(node->left);
            pre_order_traversal(node->right);
        }

        Node<T>* find(Node<T>* node, T data)
        {
            if(node == nullptr) return nullptr;

            if(data > node->data)
                return find(node->right, data);
            
            if(data < node->data)
                return find(node->left, data);

            // This means node->data == data
            return node;
        }

        int get_height(Node<T> *node)
        {
            if(node == nullptr) return 0;

            int left = get_height(node->left) + 1;
            int right = get_height(node->right) + 1;

            return left > right ? left : right;
        }
        
    public:
        AVLTree ()
        {
            this->size = 0;
            this->root = new Node<T>();
        }

        AVLTree (T data)
        {
            this->size = 1;
            this->root = new Node<T>(data);
        }
        
        int get_height()
        {
            return get_height(root);
        }

        bool contains(T data)
        {
            return find(root, data) != nullptr;
        }

        bool insert(T data)
        {
            if(data == T{})
                return false;

            // Calls recursive private insert() method.
            if(!contains(data)){
                root = insert(root, data);
                size++;
                return true;
            }

            // This is executed when we want to enter already existing data.
            return false;
        }

        T remove(T data)
        {
            if(data == T{}) return T{};

            if(size == 0) {
                cout << "The tree is empty!" << endl;
                return T{};
            }    

            if(!contains(data)){
                cout << "There is no such data" << endl;
                return T{};
            }

            remove(root, data);
            return data;
        }

        void in_order_traversal()
        {
            if(size > 0)
                in_order_traversal(root);
            else
                cout << "The tree is empty!" << endl;
            return;
        }

        void pre_order_traversal()
        {
            if(size > 0)
                pre_order_traversal(root);
            else
                cout << "The tree is empty!" << endl;
            return;
        }
};


int main(int argc, char * argv[])
{
    AVLTree<int> tree(10);
    tree.insert(20);
    tree.insert(30);
    tree.insert(40);
    tree.insert(50);
    tree.insert(60);

    //tree.remove(10);
    cout << "Height: " << tree.get_height() << endl;
    tree.pre_order_traversal();
    return 0;
}

File: BinarySearchTree/AVLTree.py
"""
    AVL Tree is Balanced Binary Tree.
    In AVL tree max height of the tree is always log(n)
    which makes the searching O(log n) even in worst case.
"""
from queue import Queue

class Node:
    def __init__(self, data) -> None:
        self.data = data
        self.left = None
        self.right = None
        self.height = 1


class AVLTree:
    def __init__(self, data=None) -> None:
        if data is not None:
            self.root = Node(data)
            self.size = 1
        else:
            self.root = None
            self.size == 0

    def _find(self, node: Node, data) -> Node:
        if node is None:
            return None
        if node.data == data:
            return node
        if data > node.data:
            return self._find(node.right, data)
        if data < node.data:
            return self._find(node.left, data)

    def _getRightSuccessor(self, node: Node) -> Node:
        successor = node.right
        while successor.left:
            successor = successor.left
        return successor

    def _getLeftSuccessor(self, node: Node) -> Node:
        successor = node.left
        while successor.right:
            successor = successor.right
        return successor

    def _getBalanceFactor(self, node: Node) -> int:
        if node is None:
            return 0
        return self.height(node.left) - self.height(node.right)
    
    def _rightRotate(self, A: Node) -> Node:
        B: Node = A.left
        Br: Node = B.right

        B.right = A
        A.left = Br

        A.height = 1 + max(self.height(A.right), self.height(A.left))
        B.height = 1 + max(self.height(B.right), self.height(B.left))        

        return B
    
    def _leftRotate(self, A: Node) -> Node:
        B: Node = A.right
        Br: Node = B.left

        B.left = A
        A.right = Br

        A.height = 1 + max(self.height(A.right), self.height(A.left))
        B.height = 1 + max(self.height(B.right), self.height(B.left))        

        return B
    
    def _updateHeight(self, node: Node) -> None:
        node.height = 1 + max(self.height(node.left), self.height(node.right))
        return node
    
    def _balanceNode(self, node: Node, data) -> Node:

        balance = self._getBalanceFactor(node)

        if balance > 1 and data < node.left.data:
            return self._rightRotate(node)
        
        if balance > 1 and data > node.left.data:
            node.left = self._leftRotate(node.left)
            return self._rightRotate(node)
        
        if balance < -1 and data < node.right.data:
            node.right = self._rightRotate(node.right)
            return self._leftRotate(node)
        
        if balance < -1 and data > node.right.data:
            return self._leftRotate(node)

        return node

    def _insert(self, node: Node, data) -> Node:
        if node is None:
            return Node(data)
        
        if data > node.data:
            node.right = self._insert(node.right, data)
        if data < node.data:
            node.left = self._insert(node.left, data)
        else:
            return node
        
        # Update height of the node.
        self._updateHeight(node)

        # Balance the node if it is unbalanced!
        return self._balanceNode(node, data)

    def _remove(self, node: Node, data) -> Node:
        if node is None:
            return None
        
        if data > node.data:
            node.right = self._remove(node.right, data)
        elif data < node.data:
            node.left = self._remove(node.left, data)
        else:
            # Here node is found!
            # Should not return here. as there is work to perform below!
            if node.left is None:
                temp = node.right
                node = temp
            elif node.right is None:
                temp = node.left
                node = temp
            else:
                # This node has two children
                successor = self._getRightSuccessor(node)
                node.data = successor.data
                node.right = self._remove(node.right, successor.data)

                # If you want to use left node instead, uncomment these lines
                # successor = self._getLeftSuccessor(data)
                # node.data = successor.data
                # self._remove(node.right, successor.data)

        # If the node is None there is no work to do.
        if node is None:
            return node
        
        # Else Perform the work to keep tree balanced
        # Work to be done to keep balance of the height among tree nodes;

        # Update the height of the node
        node = self._updateHeight(node)

        # Balance the node
        return self._balanceNode(node, data)

    def _postOrderTraversal(self, node: Node) -> None:
        if node is None:
            return 
        self._postOrderTraversal(node.left)
        self._postOrderTraversal(node.right)
        print(node.data)

    def _preOrderTraversal(self, node: Node) -> None:
        if node is None:
            return
        print(node.data)
        self._preOrderTraversal(node.left)
        self._preOrderTraversal(node.right)

    def _inOrderTraversal(self, node: Node) -> None:
        if node is None:
            return
        self._preOrderTraversal(node.left)
        print(node.data)
        self._preOrderTraversal(node.right)
    
    def getHeightOf(self, data):
        node = self._find(self.root, data)
        if node is None:
            raise KeyError("Invalid Key!")
        return self.height(node)

    def height(self, node: Node) -> int:
        if node is None:
            return 0
        return node.height
    
    def contains(self, data):
        return self._find(self.root, data) is not None

    def preOrderTraversal(self) -> None:
        if self.root is None:
            raise KeyError("Tree is empty!")
        else:
            self._preOrderTraversal(self.root)

    def postOrderTraversal(self) -> None:
        if self.root is None:
            raise KeyError("Tree is empty!")
        else:
            self._postOrderTraversal(self.root)

    def inOrderTraversal(self) -> None:
        if self.root is None:
            raise KeyError("Tree is empty!")
        else:
            self._inOrderTraversal(self.root)

    def levelOrderTraversal(self) -> None:
        if self.root is None:
            raise KeyError("The tree is empty.")
        else:
            current = self.root
            queue = Queue()
            queue.put(current)

            while not queue.empty():
                top: Node = queue.get()
                if top.left is not None:
                    queue.put(top.left)
                if top.right is not None:
                    queue.put(top.right)
                print(top.data)
            
        return
    
    def getRoot(self):
        if self.root is None:
            raise KeyError("The tree is empty.")
        else:
            return self.root.data
    
    def insert(self, data):
        if data is None:
            raise ValueError("Invalid Value!")
        self.root = self._insert(self.root, data)
        self.size = self.size + 1

    def remove(self, data):
        if not self.contains(data):
            raise KeyError("There is no key!")
        self.root = self._remove(self.root, data)
        self.size = self.size - 1


tree = AVLTree(12)

tree.insert(23)
tree.insert(10)
tree.remove(10)
tree.insert(4)
tree.insert(78)
tree.insert(14)
tree.insert(14)
tree.insert(2)
tree.insert(1)
tree.insert(29)
tree.remove(12)
tree.levelOrderTraversal()
print("Root:", tree.getRoot())
tree.preOrderTraversal()